### PR TITLE
RequirementMachine: Another silly GenericSignatureBuilder compatibility hack for concrete contraction

### PR DESCRIPTION
**Pull request by slavapestov**

### CODE REVIEW

**Comment by slavapestov** [comment]
@xedin This should fix the source compatibility test failure.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

File: test/Generics/concrete_contraction_unrelated_typealias.swift [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  typealias Parent = P
+  typealias Delegate = D
+
+  // CHECK-LABEL: .GenericDelegateProxy.init(_:)@
+  // CHECK-NEXT: <P, D, Proxy where P == Proxy.[DelegateProxyType]Parent, D == Proxy.[DelegateProxyType]Delegate, Proxy : GenericDelegateProxy<P, D>, Proxy : DelegateProxyType>
+  init<Proxy: DelegateProxyType>(_: Proxy.Type)
+    where Proxy: GenericDelegateProxy<P, D>,
+          Proxy.Parent == P,
+          Proxy.Delegate == D {}
+}
```
Comment (AnthonyLatsis) [inline test/Generics/concrete_contraction_unrelated_typealias.swift:28]: Trying to wrap my head over this. If the class is unrelated to the protocol, and the `Self` subclass could have its own distinct witnesses for these associated types, why would it be correct to substitute `GenericDelegateProxy<P, D>` into `Proxy.Parent == P` and discard the requirement?

Comment (slavapestov) [reply test/Generics/concrete_contraction_unrelated_typealias.swift:28]: I updated the test in https://github.com/apple/swift/pull/42005

**Pull request merged**